wasm_memsafety_rtlib.c

```diff
@@ -1,129 +1,103 @@
+/*
+ * We must store the size of every allocated memory region, because we need it
+ * to untag that region upon freeing.
+ *
+ * Our Strategy: For each memory allocation, we allocate an additional
+ * sizeof(size_t) bytes to store the block's size just before the actual memory
+ * block. When querying a memory region's size for freeing it, we can directly
+ * access this embedded size. This strategy results in fast memory operations at
+ * the expense of a small overhead for each allocation, but this seems in line
+ * with MTE's performance principles.
+ *
+ * The alternate approach of utilizing a hash table to track the sizes of
+ * allocated memory blocks proved to be a runtime performance bottleneck during
+ * benchmarking.
+ */
+
 #include <stdlib.h>
 #include <stdbool.h>
 #include <stdio.h>
+#include <assert.h>
 
-#define TABLE_MAX_LOAD 0.75
-
-typedef struct {
-    void *key;
-    size_t size;
-} TableEntry;
-
+#define MTE_ALIGNMENT 16
+#define MTE_NON_TAG_BITS_MASK 0xF0FFFFFFFFFFFFFF
 
 typedef struct {
-    size_t count;
-    size_t capacity;
-    TableEntry *entries;
-} HashTable;
-
-static TableEntry *__wasm_memsafety_table_find(TableEntry *entries, size_t capacity, void *ptr) {
-    size_t index = (size_t) ptr % capacity;
-    TableEntry *tombstone = NULL;
-
-    for (;;) {
-        TableEntry *entry = &entries[index];
-
-        if (entry->key == NULL) {
-            if (entry->size == 0) {
-                // found empty entry
-                return tombstone != NULL ? tombstone : entry;
-            } else if (tombstone == NULL) {
-                // found tombstone
-                tombstone = entry;
-            }
-        } else if (entry->key == ptr) {
-            return entry;
-        }
-
-        index = (index + 1) % capacity;
-    }
+  // The size of the (tagged) memory region accessible to the user. The size
+  // (of the memory) required to store this allocation metadata is excluded.
+  size_t size;
+} AllocMetadata;
```

Comment:
My idea was to store the tagged size directly in the metadata, instead of the total size like you proposed in your version. What are your thoughts on this? I think saving the tagged size saves some arithmetic operations, but maybe I am overseeing some sort of caching-related overhead/bottleneck?